tsl/src/compression/compression.c

```diff
@@ -530,6 +555,226 @@ static SegmentInfo *segment_info_new(Form_pg_attribute column_attr);
+		if (!is_null)
+		{
+			curr_seq_num = DatumGetInt32(seq_num);
+			if (max_seq_num < curr_seq_num)
+			{
+				max_seq_num = curr_seq_num;
+			}
+		}
+	}
+
+	heap_endscan(heap_scan);
+
+	return max_seq_num + SEQUENCE_NUM_GAP;
+}
+
+/* Scan compressed chunk to get the sequence number for current group.
+ * This is necessary to do when merging chunks. If the chunk is empty,
+ * scan will always return 0 and the sequence number will start from
+ * SEQUENCE_NUM_GAP.
+ */
+static int32
+get_sequence_number_for_current_group(Relation table_rel, Oid index_oid,
+									  int16 *uncompressed_col_to_compressed_col,
+									  PerColumn *per_column, int n_input_columns,
+									  int16 seq_num_column_num)
+{
+	/* No point scanning an empty relation. */
+	if (table_rel->rd_rel->relpages == 0)
+		return SEQUENCE_NUM_GAP;
+
+	/* If there is a suitable index, use index scan otherwise fallback to heap scan. */
+	bool is_index_scan = index_oid != InvalidOid;
+
+	int i, num_scankeys = 0;
+	int32 result = 0;
+
+	for (i = 0; i < n_input_columns; i++)
+	{
+		if (per_column[i].segmentby_column_index < 1)
+			continue;
+
+		num_scankeys++;
+	}
+
+	MemoryContext scan_ctx = AllocSetContextCreate(CurrentMemoryContext,
+												   "get max sequence number scan",
+												   ALLOCSET_DEFAULT_SIZES);
+	MemoryContext old_ctx;
+	old_ctx = MemoryContextSwitchTo(scan_ctx);
+
+	ScanKeyData *scankey = NULL;
+
+	if (num_scankeys > 0)
+	{
+		scankey = palloc0(sizeof(ScanKeyData) * num_scankeys);
+
+		for (i = 0; i < n_input_columns; i++)
+		{
+			if (per_column[i].segmentby_column_index < 1)
+				continue;
+
+			PerColumn col = per_column[i];
+			int16 attno =
+				is_index_scan ? col.segmentby_column_index : uncompressed_col_to_compressed_col[i];
+
+			if (col.segment_info->is_null)
+			{
+				ScanKeyEntryInitialize(&scankey[col.segmentby_column_index - 1],
+									   SK_ISNULL | SK_SEARCHNULL,
+									   attno,
+									   InvalidStrategy, /* no strategy */
+									   InvalidOid,		/* no strategy subtype */
+									   InvalidOid,		/* no collation */
+									   InvalidOid,		/* no reg proc for this */
+									   (Datum) 0);		/* constant */
+			}
+			else
+			{
+				ScanKeyEntryInitializeWithInfo(&scankey[col.segmentby_column_index - 1],
+											   0, /* flags */
+											   attno,
+											   BTEqualStrategyNumber,
+											   InvalidOid, /* No strategy subtype. */
+											   col.segment_info->collation,
+											   &col.segment_info->eq_fn,
+											   col.segment_info->val);
+			}
+		}
+	}
+
+	if (is_index_scan)
+	{
+		/* Index scan should always use at least one scan key to get the sequence number. */
+		Assert(num_scankeys > 0);
+
+		result = index_scan_sequence_number(table_rel, index_oid, scankey, num_scankeys);
+	}
+	else
+	{
+		/* Table scan can work without scan keys. */
+		result = table_scan_sequence_number(table_rel, seq_num_column_num, scankey, num_scankeys);
+	}
```

Comment:
There seems to be no test case for this situation.

Reply:
There was one but the code wasn't scanning the data properly hence not getting coverage. I've fixed the code and the test to catch it.